internal/reconciler/telemetry/status.go

```diff
@@ -0,0 +1,146 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/exp/slices"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1alpha1 "github.com/kyma-project/telemetry-manager/apis/operator/v1alpha1"
+	telemetryv1alpha1 "github.com/kyma-project/telemetry-manager/apis/telemetry/v1alpha1"
+	"github.com/kyma-project/telemetry-manager/internal/otelcollector/ports"
+	"github.com/kyma-project/telemetry-manager/internal/reconciler"
+)
+
+//go:generate mockery --name ComponentHealthChecker --filename component_health_checker.go
+type ComponentHealthChecker interface {
+	Check(ctx context.Context) (*metav1.Condition, error)
+}
+
+func (r *Reconciler) updateStatus(ctx context.Context, telemetry *operatorv1alpha1.Telemetry) error {
+	if !telemetry.GetDeletionTimestamp().IsZero() {
+		if err := r.updateStateIfBeingDeleted(ctx, telemetry); err != nil {
+			return fmt.Errorf("failed to check if telemetry is being deleted: %w", err)
+		}
+
+		// Interrupt further state evaluation if the resource is being deleted
+		return nil
+	}
+
+	for _, checker := range r.enabledHealthCheckers() {
+		if err := r.updateComponentCondition(ctx, checker, telemetry); err != nil {
+			return fmt.Errorf("failed to update component condition: %w", err)
+		}
+	}
+
+	if err := r.updateGatewayEndpoints(ctx, telemetry); err != nil {
+		return fmt.Errorf("failed to update gateway endpoints: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) enabledHealthCheckers() []ComponentHealthChecker {
+	if r.config.Metrics.Enabled {
+		return []ComponentHealthChecker{r.healthCheckers.logs, r.healthCheckers.metrics, r.healthCheckers.traces}
+	}
+	return []ComponentHealthChecker{r.healthCheckers.logs, r.healthCheckers.traces}
+}
+
+func (r *Reconciler) updateComponentCondition(ctx context.Context, checker ComponentHealthChecker, telemetry *operatorv1alpha1.Telemetry) error {
+	newCondition, err := checker.Check(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to check component: %w", err)
+	}
+
+	newCondition.ObservedGeneration = telemetry.GetGeneration()
+	meta.SetStatusCondition(&telemetry.Status.Conditions, *newCondition)
+
+	if slices.ContainsFunc(telemetry.Status.Conditions, func(cond metav1.Condition) bool {
+		return cond.Status == metav1.ConditionFalse
```

Comment:
The [api-conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) contain the following description about the condition status:

```
For some conditions, True represents normal operation, and for some conditions, False represents normal operation. ("Normal-true" conditions are sometimes said to have "positive polarity", and "normal-false" conditions are said to have "negative polarity".) Without further knowledge of the conditions, it is not possible to compute a generic summary of the conditions on a resource.
```

So this cannot be a generic implementation. Since all our current conditions have "positive polarity", the implementation is correct. But maybe let's add a comment with a hint.

Reply:
Good idea, will do